### PR TITLE
Fix issue #9858.

### PR DESCRIPTION
**Pull request by jiefangxuanyan**
Added some byte swapping code to solve endianness problem.

### CODE REVIEW

**Comment by jiefangxuanyan** [comment]
I signed it!

File: tensorflow/core/kernels/decode_raw_op.cc [hunk 1]
```diff
@@ -73,6 +73,13 @@ class DecodeRawOp : public OpKernel {
     for (int64 i = 0; i < flat_in.size(); ++i) {
       const T* in_data = reinterpret_cast<const T*>(flat_in(i).data());
       memcpy(out_data, in_data, str_size);
+      if (port::kLittleEndian != little_endian_) {
```
Comment (rmlarsen) [inline tensorflow/core/kernels/decode_raw_op.cc]: Instead if memcpy + byteswap, just use a custom copy loop for the big endian case, so you don't have to touch the data twice.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Comment by jiefangxuanyan** [comment]
@rmlarsen like this?

**Review by rmlarsen — APPROVED** [review]

File: tensorflow/core/kernels/decode_raw_op.cc [hunk 2]
```diff
@@ -70,10 +70,24 @@ class DecodeRawOp : public OpKernel {
...
-      const T* in_data = reinterpret_cast<const T*>(flat_in(i).data());
-      memcpy(out_data, in_data, str_size);
-      out_data += added_dim;
+    if (port::kLittleEndian == little_endian_) {
+      for (int64 i = 0; i < flat_in.size(); ++i) {
+        const T* in_data = reinterpret_cast<const T*>(flat_in(i).data());
+        memcpy(out_data, in_data, str_size);
+        out_data += added_dim;
+      }
+    } else {
```
Comment (rmlarsen) [inline tensorflow/core/kernels/decode_raw_op.cc]: Please add unit test(s) such that both branches are tested.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
Please add unit test(s).

**Review by rmlarsen — CHANGES_REQUESTED** [review]
Please add a unit test.

**Comment by jiefangxuanyan** [comment]
I added a test case for endianness. By the way, according to the documentation (https://www.tensorflow.org/versions/master/api_docs/python/tf/decode_raw), I think this op should behave consistently on platforms with different endianness, so I removed the endianness test in an existing test case. If the original behavior in that test case is expected, maybe you should make changes on the documentation.

File: tensorflow/python/kernel_tests/decode_raw_op_test.py [hunk 3]
```diff
@@ -54,18 +54,26 @@ def testToInt16(self):
...
         decode.eval(feed_dict={in_bytes: ["123", "456"]})
 
+  def testEndianness(self):
+    with self.test_session():
+      in_bytes = array_ops.placeholder(dtypes.string, shape=[None])
+      decode_le = parsing_ops.decode_raw(
+          in_bytes, out_type=dtypes.int32, little_endian=True)
+      decode_be = parsing_ops.decode_raw(
+          in_bytes, out_type=dtypes.int32, little_endian=False)
+      result = decode_le.eval(feed_dict={in_bytes: ["\x01\x02\x03\x04"]})
```
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/decode_raw_op_test.py]: Won't this depend on the endianness of the machine where this test is run? You can test the endianness of the machine here using the Python sys module, e.g.,
```
>>> import sys
>>> sys.byteorder
'little'
```

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/decode_raw_op_test.py (hunk 3 above)
Comment (jiefangxuanyan) [inline tensorflow/python/kernel_tests/decode_raw_op_test.py]: It won't. This decode op gets same result on machines with different endianness, just like the way these machines receive data from the network. In fact, platform endianness is considered in the C++ code (`if (port::kLittleEndian == little_endian_)`) to ensure consistency. As I said before, the old platform-dependent behavior in this test is contradictionary with the documentation.

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Pull request merged**